### PR TITLE
combine flatten and hash

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
Scanning accounts and accumulating hash and lamports is expensive. Flattening large vectors of data (from individual elements or from sub-vectors) is also expensive.

#### Summary of Changes
Allow hash calculation to operate on vectors of smaller size, resulting in lower time and memory use.
Later, we can do partial hashing of results to further reduce the data that has to be retained in memory during a scan.

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
This pr is related to #15275. After that pr is ready, we will be able to reduce some code paths and remove some redundancy. There are 2 separate ideas, so I broke it into 2 PRs.

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -3813,14 +3953,31 @@ impl AccountsDB {
...
+    fn flatten_hashes_and_hash(
+        hashes: Vec<Vec<Hash>>,
+        fanout: usize,
+        test: bool,
+    ) -> (Hash, Measure, usize) {
         let mut flat2_time = Measure::start("flat2");
-        let hashes: Vec<Hash> = hashes.into_iter().flatten().collect();
+
+        let offsets = CumulativeOffsets1D::from_raw(&hashes);
+        let hash_bkup = if test {
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:3964]: I would have this path just in the test code itself. We should be able to call both paths and compare in a test.

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -190,6 +190,64 @@ impl CalculateHashIntermediate {
...
+
+impl CumulativeOffsets1D {
+    pub fn from_raw<T>(raw: &[Vec<T>]) -> CumulativeOffsets1D {
+        let mut total_count: usize = 0;
+        let mut cumulative_offsets = Vec::new();
+        raw.iter().enumerate().for_each(|(i, v)| {
+            let len = v.len();
+            cumulative_offsets.push(CumulativeOffset1D::new(i, total_count, total_count + len));
+            total_count += len;
+        });
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:226]: seems like a natural collect
`let cumulative_offsets = raw.iter().enumerate().map(|(i, v)| ...).collect();`

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (jeffwashington) [reply runtime/src/accounts_db.rs]: it was already also being tested in tests

File: runtime/src/accounts_db.rs (hunk 2 above)
Comment (jeffwashington) [reply runtime/src/accounts_db.rs]: yes, and it would be good to remove zero length vectors here, so did that, too

**Review by sakridge — DISMISSED** [review]

**Pull request merged**